File: Python/omniPy/FileSystem/winReg_QueryValue.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys, os, re, winreg

def winReg_QueryValue( key, val_name = None ) -> 'Get the value of [val_name] within the [key] of Windows Registry':
    #000. Info.
    '''
#---------------------------------------------------------------------------------------------------------------------------------------#
#100.   Introduction.                                                                                                                   #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |This function is intended to query the value of [val_name] within the [key] of Windows Registry.                                   #
#   |It is useful to search for the installation path of any specific software on current Windows OS                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#200.   Glossary.                                                                                                                       #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Parameters.                                                                                                                 #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |key         :   Full valid path for Windows Registry within which to query the value                                               #
#   |val_name    :   Name of the sub-key within current [key], for which to query the value                                             #
#   |                 [None        ] <Default> Retrieve the [Default Value] of current [key], as indicated [(Default)] in Registry      #
#   |                 [<str>       ]           Provide a sub-key for query                                                              #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |900.   Return Values by position.                                                                                                  #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |<various>   :   The search result in various types with respect of the Windows Registry. [None] is returned if not found           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#300.   Update log.                                                                                                                     #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   | Date |    20220117        | Version | 1.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |Version 1.                                                                                                                  #
#   |______|____________________________________________________________________________________________________________________________#
#---------------------------------------------------------------------------------------------------------------------------------------#
#400.   User Manual.                                                                                                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |See the [Full Test Program] section                                                                                                #
#---------------------------------------------------------------------------------------------------------------------------------------#
#500.   Dependent Facilities.                                                                                                           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Dependent Modules                                                                                                           #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |sys, os, re, winreg                                                                                                            #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |300.   Dependent user-defined functions                                                                                            #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------------#
    '''

    #001. Import necessary functions for processing.

    #010. Check parameters.
    #011. Prepare log text.
    #python 动态获取当前运行的类名和函数名的方法: https://www.cnblogs.com/paranoia/p/6196859.html
    LfuncName : str = sys._getframe().f_code.co_name

    #012. Handle the parameter buffer.
    if len(key) == 0: return(None)

    #050. Local parameters
    fullPath, rootName = os.path.split(key)
    while len(fullPath):
        fullPath, rootKey = os.path.split(fullPath)

    keyPath = re.sub(r'^[\\/]+', '', key[len(rootKey):])

    #100. Setup the connection to Windows Registry
    regConn = winreg.ConnectRegistry(None, winreg.__getattribute__(rootKey))

    #500. Try to obtain the value for the requested key
    #[ASSUMPTION]:
    #[1] The provided [key] may be invalid
    #[2] There may not be a [sub-key] of the provided [key]
    try:
        accKey = winreg.OpenKey(regConn, keyPath)
        rstOut = winreg.QueryValueEx(accKey, val_name)[0]
    except:
        rstOut = None

    #700. Close the connection
    winreg.CloseKey(regConn)

    #999. Return the values
    return(rstOut)
#End winReg_QueryValue

'''
#-Notes- -Begin-
#Full Test Program[1]:
if __name__=='__main__':
    #010.   Create envionment.
    import sys
    dir_omniPy : str = r'D:\Python\ '.strip()
    if dir_omniPy not in sys.path:
        sys.path.append( dir_omniPy )
    from omniPy.FileSystem import winReg_QueryValue
    print(winReg_QueryValue.__doc__)

    #100. Retrieve [SAS] installation path
    sasVer = '9.4'
    sasKey = os.path.join(r'HKEY_LOCAL_MACHINE\SOFTWARE\SAS Institute Inc.\The SAS System', sasVer)
    sasVal = r'DefaultRoot'
    print(winReg_QueryValue(sasKey, sasVal))

    #200. Retrieve [Python] installation path
    pyVer = '3.7'
    pyKey = os.path.join(r'HKEY_LOCAL_MACHINE\SOFTWARE\Python\PythonCore', pyVer, 'InstallPath')
    pyVal = None
    print(winReg_QueryValue(pyKey, pyVal))

    #300. Retrieve [R] installation path
    rKey = r'HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64'
    rVal = r'InstallPath'
    print(winReg_QueryValue(rKey, rVal))
#-Notes- -End-
'''

File: R/omniR/FileSystem/winReg_QueryValue.r
#---------------------------------------------------------------------------------------------------------------------------------------#
#100.   Introduction.                                                                                                                   #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |This function is intended to query the value of [val_name] within the [key] of Windows Registry.                                   #
#   |It is useful to search for the installation path of any specific software on current Windows OS                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#200.   Glossary.                                                                                                                       #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Parameters.                                                                                                                 #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |key         :   Full valid path for Windows Registry within which to query the value                                               #
#   |val_name    :   Name of the sub-key within current [key], for which to query the value                                             #
#   |                 [<missing>   ] <Default> Retrieve the [Default Value] of current [key], as indicated [(Default)] in Registry      #
#   |                 [<str>       ]           Provide a sub-key for query                                                              #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |900.   Return Values by position.                                                                                                  #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |<str>       :   Character vector as query result                                                                                   #
#---------------------------------------------------------------------------------------------------------------------------------------#
#300.   Update log.                                                                                                                     #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   | Date |    20220213        | Version | 1.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |Version 1.                                                                                                                  #
#   |______|____________________________________________________________________________________________________________________________#
#---------------------------------------------------------------------------------------------------------------------------------------#
#400.   User Manual.                                                                                                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |See the [Full Test Program] section                                                                                                #
#---------------------------------------------------------------------------------------------------------------------------------------#
#500.   Dependent Facilities.                                                                                                           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Dependent Modules                                                                                                           #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |base                                                                                                                           #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |300.   Dependent user-defined functions                                                                                            #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------------#

winReg_QueryValue <- function(key, val_name){
	#001. Handle parameters
	#[Quote: https://stackoverflow.com/questions/15595478/how-to-get-the-name-of-the-calling-function-inside-the-called-routine ]
	LfuncName <- deparse(sys.call()[[1]])
	#If above statement cannot find the name correctly, this function must have been called via [do.call] or else,
	# hence we need to traverse one layer above current one and extract the first argument of that call.
	if (grepl('^function.+$',LfuncName[[1]],perl = T)) LfuncName <- gsub('^.+?\\((.+?),.+$','\\1',deparse(sys.call(-1)),perl = T)[[1]]

	#012. Parameter buffer
	if (missing(key)) return(NULL)
	if (length(key) == 0) return(character(0))
	if (missing(val_name)) val_name <- NA

	#200. Define helper functions
	#210. Function to process each element of the input vectors
	proc_query <- function(key, name){
		#100. Define the switch for the Windows Command [REG QUERY]
		if (length(name) == 0) name <- NA
		if (is.na(name)) {
			reg_switch <- '/ve'
			name <- '\\(.+?\\)'
		} else {
			reg_switch <- paste('/v', shQuote(name))
		}

		#200. Prepare the query command
		reg_cmd <- paste('REG', 'QUERY', shQuote(key), reg_switch)

		#300. Query the key from Windows Registry and return a list of characters
		reg_rst <- suppressWarnings(system(reg_cmd, intern = T))

		#399. Return NA if the query command fails
		reg_status <- attr(reg_rst, 'status')
		if (length(reg_status)) if (reg_status != 0) {return(NA)}

		#500. Define the regular expression to match the query result
		reg_exp <- paste0('^\\s*', name, '\\s{4}REG_\\w+\\s{4}(.+)\\s*$')

		#700. Clean the result from direct query
		reg_cln <- sapply(
			reg_rst
			,function(x){
				if (grepl(reg_exp, x, ignore.case = T, perl = T)) {
					return(gsub(reg_exp, '\\1', x, ignore.case = T, perl = T))
				}
			}
			,USE.NAMES = F
		)
		reg_cln <- unlist(Filter(Negate(is.null), reg_cln))

		#790. Set the default result as NA
		if (length(reg_cln) == 0) reg_cln <- NA

		#999. Return the cleaned result
		return(reg_cln)
	}

	#999. Apply the helper function to the input vectors
	return(unlist(mapply(proc_query, key, val_name, SIMPLIFY = T, USE.NAMES = F)))
}

#[Full Test Program;]
if (FALSE){
	#Simple test
	if (TRUE){
		#100. Prepare strings
		reg_paths <- c(
			'HKEY_LOCAL_MACHINE\\SOFTWARE\\SAS Institute Inc.\\The SAS System\\9.4'
			,'HKEY_LOCAL_MACHINE\\SOFTWARE\\Python\\PythonCore\\3.7\\InstallPath'
			,'HKEY_LOCAL_MACHINE\\SOFTWARE\\R-core\\R64'
		)
		reg_names <- c(
			'DefaultRoot'
			,NA
			,'InstallPath'
		)

		#200. Test query
		reg_rst <- winReg_QueryValue(reg_paths, reg_names)

		#300. Retrieve the version of current Windows OS
		#Quote: https://mivilisnet.wordpress.com/2020/02/04/how-to-find-the-windows-version-using-registry/
		winver <- winReg_QueryValue('HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion', 'CurrentVersion')

		#400. Retrieve the <Language for non-Unicode Programs> of current Windows OS
		#[ASSUMPTION]
		#[1] In case there is a pre-defined key <Default> (rather than <(Default)>), we obtain its value
		#[2] Otherwise we obtain the <(Default)> value during OS installation
		#Quote: https://serverfault.com/questions/957167/windows-10-1809-region-language-registry-keys
		get_locale <- winReg_QueryValue('HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\Nls\\Locale', 'Default')
		if (is.na(get_locale)) {
			get_locale <- winReg_QueryValue('HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\Nls\\Locale')
		}
		message(get_locale)

		#500. Test calculation in [dplyr]
		library(magrittr)
		df_reg <- data.frame(x = reg_paths, y = reg_names, stringsAsFactors = F) %>%
			dplyr::mutate(p = winReg_QueryValue(x,y))
	}
}
